lib/libpcsc-cpp/include/pcsc-cpp/pcsc-cpp.hpp

```diff
@@ -142,52 +142,37 @@ struct ResponseApdu
 /** Struct that wraps command APDUs. */
 struct CommandApdu
 {
-    byte_type cla;
-    byte_type ins;
-    byte_type p1;
-    byte_type p2;
-    unsigned short le;
-    // Lc is data.size()
-    byte_vector data;
+    static constexpr size_t MAX_DATA_SIZE = 255;
 
-    static const size_t MAX_DATA_SIZE = 255;
-    static const unsigned short LE_UNUSED = std::numeric_limits<unsigned short>::max();
+    // Case 1
+    CommandApdu(byte_type cls, byte_type ins, byte_type p1, byte_type p2) : d {cls, ins, p1, p2} {}
 
-    CommandApdu(byte_type c, byte_type i, byte_type pp1, byte_type pp2, byte_vector d = {},
-                unsigned short l = LE_UNUSED) :
-        cla(c), ins(i), p1(pp1), p2(pp2), le(l), data(std::move(d))
+    // Case 2
+    CommandApdu(byte_type cls, byte_type ins, byte_type p1, byte_type p2, byte_type le) :
+        d {cls, ins, p1, p2, le}
     {
     }
 
-    CommandApdu(const CommandApdu& other, byte_vector d) :
-        CommandApdu(other.cla, other.ins, other.p1, other.p2, std::move(d), other.le)
+    // Case 3
+    CommandApdu(byte_type cls, byte_type ins, byte_type p1, byte_type p2, byte_vector data) :
+        d {std::move(data)}
     {
+        if (d.size() > MAX_DATA_SIZE) {
+            throw std::invalid_argument("Command chaining and extended lenght not supported");
+        }
+        d.insert(d.begin(), {cls, ins, p1, p2, static_cast<byte_type>(d.size())});
```

Comment:
insert() can cause a memory copy.

Let's use make_move_iterator for moving the data?

Reply:
how make_move_iterator avoids vector resize? it does same way resize to current vector and moves elements from one vector to other.

Reply:
make_move_iterator does cannot avoid vector resize, but it can avoid a hidden memory copy if we allocate a sufficiently large target vector and move elements from the source vector to the target vector.